### DISCUSSION TITLE
It is possible to create a PreRun timeout script?

### DISCUSSION THREAD

**Discussion by jighen555** [thunderclient/thunder-client-support#957] [Q&A]
When i run an HTTP call, i need to sleep 10 seconds after the API return a response.

It is possible to create a PreRun timeout script? 
Something like this (Postman example):
  function sleep() {
      setTimeout(function(){
          console.log("Sleeping for 10 seconds");
      }, 10000);
  }

**Comment by rangav** [comment]
Why you like to use sleep for 10s after response, just trying to understand your use case?

Are you running multiple pre requests?

**Comment by jighen555** [comment]
Yes exactly, i have to run two call and i need to wait 10 second when a team member run the second call.
I need to implement this security requirement for my company.

**Reply by rangav** [reply]
Yes you should able to sleep, better to use sleep in pre-request than post request.

please test and let me know if you have any issues.

**Reply by jighen555** [reply]
Unfortunately i try without success to include this script:
  function sleep() {
      setTimeout(function(){
          console.log("Sleeping for 40 seconds");
      }, 40000);
  }
  module.exports = [sleep];

![image](https://user-images.githubusercontent.com/104869891/210243396-e264f838-b9eb-4415-9e5b-9a607ffdd52f.png)

The HTTP call not run a sleeping time before send the request.
Am i doing something wrong?

Thank you very much for answer the second day of the year :-)

**Reply by rangav** [reply]
Happy New year, you are welcome.

you should use a async version of sleep.

 ```
async function sleep(){

await new Promise(res => setTimeout(res, 40000)); 

}
```

**Reply by jighen555** [reply]
I'm so sorry, with this script not work:
###  
async function sleep(){
      await new Promise(res => setTimeout(res, 40000)); 
  }
  module.exports = [sleep];

Please put the script

**Reply by rangav** [reply]
Yes the above should work

**Reply by jighen555** [reply]
For me not work, if you find another solution please put in the comments.

**Reply by rangav** [reply]
Will test and get back to you tomorrow

**Reply by jighen555** [reply]
Thank you very much!

**Reply by rangav** [reply]
`setTimeout` function is not enabled in scripts yet, Will enable in the next update.

**Reply by jighen555** [reply]
Ok good! When about the update will be available?
Consider that a big world company want to replace postman with your software, it could be a good advertisement for you and your software.

Thank you for the support!

**Reply by rangav** [reply]
that’s great to know.

Planning to release new update next week.

Can I know your company name and no of devs planning to use the extension?

**Reply by jighen555** [reply]
Unfortunately I am not able to put the name of my company for privacy, I am so sorry, but when we will terminate our migration from Postman to Thunderclient I can say you the name without problems 👍

**Reply by rangav** [reply]
Sure, No problem

**Comment by rangav** [comment]
This feature is now implemented and published to marketplace, please update to `v2.3.0`

See all features released
https://github.com/rangav/thunder-client-support/releases/tag/v2.3.0

Please let me know your feedback.

**Comment by jighen555** [comment]
I tested the sleep time and it works!
Thank you very very much for this super fix developed in such short time.

I follow you on linkedin, if you will accept me i want to ask some other question, if it possible. 
Otherwise, good luck for the future!

**Reply by rangav** [reply]
Thanks for confirmation, Sure let me know your question

**Reply by jighen555** [reply]
Now i understand why execute scripts before the rest call.
It is possible to run the sleep time script after the rest call? For example in the Tests tab:
![image](https://user-images.githubusercontent.com/104869891/210855781-5ca13c9f-f565-4e84-8e09-4c240520ba80.png)

setTo is just an example, my finally goal is only run the sleep time script after the HTTP call and before return the response.
Basically delay the response.

**Reply by rangav** [reply]
You should be able to use in Tests

steps here
https://github.com/rangav/thunder-client-support/blob/master/docs/filters.md#filters-in-tests

1. select `Json Query` test

2. Use `json | sleep` in the left input

**Comment by jighen555** [comment]
Sorry i have another question regarding the sleeptime.
Now my flow is:
Run HTTP call --> wait 40 seconds --> return response and console.log("Sleep time 40 sec")

It is possibile to create this flow?
Run HTTP call --> return response --> wait 40 seconds --> console.log("Sleep time 40 sec")

**Reply by rangav** [reply]
You can use the custom filter in Tests and have the sleep functionality.

reference
https://github.com/rangav/thunder-client-support/discussions/957#discussioncomment-4604697
https://github.com/rangav/thunder-client-support/discussions/972#discussioncomment-4665914

**Reply by rangav** [reply]
If you like run multiple request in `Run Collection`, there is delay input field available

<img width="1172" alt="Screenshot 2023-01-12 at 11 42 59" src="https://user-images.githubusercontent.com/8637550/212058364-0977801a-3f62-4a38-80c5-2afd649dae2a.png">
